### PR TITLE
Make wait-for-ready Documentation Clearer

### PR DESCRIPTION
**Pull request by ncteisen**
This concept has caused much confusion, especially with regard to retries. Here is an attempt, at the risk of over explaining, to make it clearer. I've added everyone from our offline discussion plus others to ensure we are on the same page here.

### CODE REVIEW

File: doc/wait-for-ready.md [hunk 1]
```diff
@@ -1,14 +1,26 @@
...
-gRPC implementations MAY provide a per-RPC option to not fail RPCs as a result
-of the channel being in `TRANSIENT_FAILURE` state. Instead, the implementation
-queues the RPCs until the channel is `READY`. This is known as "wait for ready."
-The RPCs SHOULD still fail before `READY` if there are unrelated reasons, such
-as the channel is `SHUTDOWN` or the RPC's deadline is reached.
+Errors can occur in many places. For the propose of this document we can bucket the errors into three separate situations:
+
+1. The RPC never leaves the gRPC client library, for example:
+    * The connection was broken and the channel is in a `TRANSIENT_FAILURE` state.
+2. The RPC reaches the gRPC server application, and is rejected. For example:
```
Comment (markdroth) [inline doc/wait-for-ready.md]: I think this should say server *library* here, not server *application*, right?

File: doc/wait-for-ready.md [hunk 2]
```diff
@@ -1,14 +1,26 @@
...
+    * The server is momentarily overloaded and sends a GOAWAY frame.
+3. The RPC reaches the server application, and fails.
+
+![Where RPCs Fail](images/WhereRPCsFail.png)
+
+In all of these cases, by default, gRPC implementations SHOULD immediately return the error to the client application. This is known as "fail fast" but the usage of the term is historical.
+
+gRPC implementations MAY provide a per-RPC option to not fail RPCs as a result of reason 1 and 2 (the cases in which the RPC never made it to the server application). This is known as "wait for ready" behavior. an RPC with this behavior enabled will stay pending on the client until:
```
Comment (markdroth) [inline doc/wait-for-ready.md]: Need to capitalize "an" at the start of the last sentence.

Comment (ncteisen) [reply doc/wait-for-ready.md]: Done

File: doc/wait-for-ready.md (hunk 1 above)
Comment (ncteisen) [reply doc/wait-for-ready.md]: Yes it should! Good catch

**Comment by ncteisen** [comment]
I'll wait a day to give others a chance to comment then merge.

**Comment by markdroth** [comment]
I suggest getting positive ack from @a11r before merging.

File: doc/wait-for-ready.md [hunk 3]
```diff
@@ -1,14 +1,26 @@
...
-implementations SHOULD fail such RPCs immediately. This is known as "fail fast,"
-but usage of the term is historical. RPCs SHOULD NOT fail as a result of the
-channel being in other states (`CONNECTING`, `READY`, or `IDLE`).
-
-gRPC implementations MAY provide a per-RPC option to not fail RPCs as a result
-of the channel being in `TRANSIENT_FAILURE` state. Instead, the implementation
-queues the RPCs until the channel is `READY`. This is known as "wait for ready."
-The RPCs SHOULD still fail before `READY` if there are unrelated reasons, such
-as the channel is `SHUTDOWN` or the RPC's deadline is reached.
+Errors can occur in many places. For the propose of this document we can bucket the errors into three separate situations:
```
Comment (ejona86) [inline doc/wait-for-ready.md]: Please reflow text to 80 columns.

**Comment by dfawley** [comment]
Thanks for writing the clarifications.  Would it be possible to squeeze in a mention that the status returned is UNAVAILABLE for transient failures with fail-fast (like #12390 is attempting to do).

Also, I'd change "MUST be cautious" into "should be cautious".  "Being cautious" is subjective and not specific, so it's pretty hard to say if a gRPC implementation is cautious.  If we can define a specific set of requirements here, using "MUST" would be appropriate.  E.g. "gRPC implementations MUST NOT make more than three attempts that could reach the [or 'a single'?] remote server".

File: doc/wait-for-ready.md [hunk 4]
```diff
@@ -1,14 +1,26 @@
...
+    * The connection was broken and the channel is in a `TRANSIENT_FAILURE` state.
+2. The RPC reaches the gRPC server library, and is rejected. For example:
+    * The server is momentarily overloaded and sends a GOAWAY frame.
+3. The RPC reaches the server application, and fails.
+
+![Where RPCs Fail](images/WhereRPCsFail.png)
+
+In all of these cases, by default, gRPC implementations SHOULD immediately return the error to the client application. This is known as "fail fast" but the usage of the term is historical.
+
+gRPC implementations MAY provide a per-RPC option to not fail RPCs as a result of reason 1 and 2 (the cases in which the RPC never made it to the server application). This is known as "wait for ready" behavior. An RPC with this behavior enabled will stay pending on the client until:
```
Comment (markdroth) [inline doc/wait-for-ready.md]: Suggest adding the word "either" at the end of this line, just to make the meaning explicitly clear.

File: doc/wait-for-ready.md [hunk 5]
```diff
@@ -1,14 +1,26 @@
...
+  * it sees an *irrecoverable* failure
+  * its deadline is reached
+
+Recalling the examples at top:
+  * an overloaded server may recover
+  * a channel in `TRANSIENT_FAILURE` may become connected
+
+So we do not consider these types of failures *irrecoverable*.
+
+**NOTE**: error case 2 needs one more point of clarification. gRPC implementations that support "wait for ready" MUST be cautious to protect the server when the failure occurs in the gRPC server library. Transparently retrying indefinitely in this case will burden the server and increase wire traffic. In these cases implementations should make a reasonable attempt to "wait for ready" (like retrying with backoff three times), then fail back to the client application.
```
Comment (ejona86) [inline doc/wait-for-ready.md]: This sounds like you're actually changing the spec. I don't see this case covered at all in the retry spec. Is this the first this has been discussed?

File: doc/wait-for-ready.md [hunk 6]
```diff
@@ -1,14 +1,26 @@
...
-
-gRPC implementations MAY provide a per-RPC option to not fail RPCs as a result
-of the channel being in `TRANSIENT_FAILURE` state. Instead, the implementation
-queues the RPCs until the channel is `READY`. This is known as "wait for ready."
-The RPCs SHOULD still fail before `READY` if there are unrelated reasons, such
-as the channel is `SHUTDOWN` or the RPC's deadline is reached.
+Errors can occur in many places. For the propose of this document we can bucket the errors into three separate situations:
+
+1. The RPC never leaves the gRPC client library, for example:
+    * The connection was broken and the channel is in a `TRANSIENT_FAILURE` state.
```
Comment (ejona86) [inline doc/wait-for-ready.md]: "The connection was broken" does not trigger "TRANSIENT_FAILURE state". "The connection was broken" triggers IDLE. It sounds like you just want "the channel is in a `TRANSIENT_FAILURE` state".

File: doc/wait-for-ready.md (hunk 5 above)
Comment (ejona86) [inline doc/wait-for-ready.md]: Rejecting a stream does _not_ imply backoff. It should be replayed _immediately_.

File: doc/wait-for-ready.md [hunk 7]
```diff
@@ -1,14 +1,26 @@
...
+2. The RPC reaches the gRPC server library, and is rejected. For example:
```
Comment (ejona86) [inline doc/wait-for-ready.md]: Bringing this into things is re-defining, not "clarifying". I think this is mostly orthogonal to wait-for-ready/fail-fast. "fail fast" should be permitted to be replayed here. Now, we may choose not to, but I don't think we should permanently disallow it. We can discuss this particular case somewhere else; it doesn't need to be here as it has nothing to do with the semantics of either fail fast or wait-for-ready.

**Comment by ejona86** [comment]
> Would it be possible to squeeze in a mention that the status returned is UNAVAILABLE for transient failures with fail-fast

That seems unrelated to this document. It shouldn't talk about status codes at all.

**Comment by dfawley** [comment]
> That seems unrelated to this document. It shouldn't talk about status codes at all.

How isn't it related?  This is the only place that discusses "fail fast", and this status code is returned as a result of a fail fast RPC encountering a transient failure.

File: doc/wait-for-ready.md [hunk 8]
```diff
@@ -1,14 +1,26 @@
...
+![Where RPCs Fail](images/WhereRPCsFail.png)
+
+In all of these cases, by default, gRPC implementations SHOULD immediately return the error to the client application. This is known as "fail fast" but the usage of the term is historical.
+
+gRPC implementations MAY provide a per-RPC option to not fail RPCs as a result of reason 1 and 2 (the cases in which the RPC never made it to the server application). This is known as "wait for ready" behavior. An RPC with this behavior enabled will stay pending on the client until:
+  * it sees an *irrecoverable* failure
+  * its deadline is reached
+
+Recalling the examples at top:
+  * an overloaded server may recover
```
Comment (a11r) [inline doc/wait-for-ready.md]: Let's not talk about "why" the server is doing something. Instead, this example should be reworked to talk about what the client might do when it receives a GOAWAY. (i.e., the client might be able to get a new working connection ).

**Comment by ejona86** [comment]
> How isn't it related? This is the only place that discusses "fail fast", and this status code is returned as a result of a fail fast RPC encountering a transient failure.

This does not _define_ fail fast. This document only discussed it to contrast. Yes, this is the only place that mentions the term "fail fast", but that's in part because the term _does not exist_ in gRPC. We only have "wait for ready". The "default behavior" (sometimes called "fail fast") should be defined elsewhere.

File: doc/wait-for-ready.md (hunk 4 above)
Comment (ncteisen) [reply doc/wait-for-ready.md]: Resolved

**Comment by dfawley** [comment]
> The "default behavior" (sometimes called "fail fast") should be defined elsewhere.

This document _does_ define the default behavior:

Before:
> By default, gRPC implementations SHOULD fail such RPCs immediately.

After:
> by default, gRPC implementations SHOULD immediately return the error to the client application

Until a separate document exists that more thoroughly defines the default behavior, I think it's perfectly reasonable to add one extra detail to the behavior that is described here.

**Comment by ejona86** [comment]
> Until a separate document exists that more thoroughly defines the default behavior

https://github.com/grpc/grpc/pull/12434

Wait for ready is an optional feature. It should not be a place for defining required features. I'd be okay if the SHOULDs in the first paragraph were removed so that they simply stated the state of the world.

**Comment by ejona86** [comment]
After #12434 this document wouldn't actually need to define anything new.

File: doc/wait-for-ready.md (hunk 3 above)
Comment (ncteisen) [reply doc/wait-for-ready.md]: Done

File: doc/wait-for-ready.md (hunk 6 above)
Comment (ncteisen) [reply doc/wait-for-ready.md]: Due to a know c core issue, a broken connection can lead to TRANSIENT_FAILURE in some cases. However, I realize this is an implementation detail, and should be removed from the doc. I have removed specific examples to more general: If an RPC never leaves the client, it belongs in this bucket.

File: doc/wait-for-ready.md (hunk 7 above)
Comment (ncteisen) [reply doc/wait-for-ready.md]: To be discussed

File: doc/wait-for-ready.md (hunk 5 above)
Comment (ncteisen) [reply doc/wait-for-ready.md]: To be discussed

Comment (ncteisen) [reply doc/wait-for-ready.md]: To be discussed

**Comment by ncteisen** [comment]
Updated the doc to be a little bit more general, removing specific examples, and clarifying things. We can iron out all the disagreed-upon details offline today

File: doc/wait-for-ready.md (hunk 6 above)
Comment (ejona86) [reply doc/wait-for-ready.md]: Also: I was wrong, TRANSIENT_FAILURE can be triggered by broken connections.

**Comment by ncteisen** [comment]
Just did a run over old PRs. Is this semantic update good? Should we continue this thread and discuss? Should I drop this?

**Comment by ejona86** [comment]
@ncteisen I don't think this represents our current understanding. I thought we decided FailFast/WaitForReady would not impact cases 1 and 2; they would always be done because there's no sleeps involved (we're still making progress).

**Pull request closed without merging**